### PR TITLE
Add ArtificialExtendsBoundTree to solve the wildcard issue.

### PR DESCRIPTION
**Pull request by CharlesZ-Chen**
Originally, `VariableAnnotator` would only add one VarAnno on an unbounded wildcard type, which means its super bound and extend bound share the same VarAnno. This would cause conflict constraints being generated in some situation:

```
// given wildcard: @VarAnno(1) ? extends @VarAnno(1) Object,
// when assign a wildcard type with default annotations to above wildcard: 
// @BottomRealQual ? extends @TopRealQual Object
// below constraint generated:

@VarAnno(1) <: BottomRealQual
@VarAnno(1) :> TopRealQual
```

A proper solution would be use two different VarAnnos to represent super bound and extend bound for a given unbounded wildcard. If we infer the super bound as Bottom, and the extend bound as Top, then we just silently ignore those solutions and don't insert them into the source code (as this is equal to original unbounded wildcard). Otherwise we either insert the solution or notifying user with the solution to suggest different bounds for that wildcard.).

To achieve above solution, we need create a fake extend bound tree for unbounded wildcard type in order to enable `VariableAnnotator` insert different VarAnnos on its bounds.

Below is the changes in this PR:

- Add `ArtificialExtendsBoundTree` to represents an artificial extends bound tree for an unbounded wildcard type.
- Add `ArtificialTreeBuilder` as Tree factory to build artificial trees.

- Change the logic of `VariableAnnotator#visitWildcard()`, to add different VarAnnos on extends bound and super bound for a given unbounded wildcard type.
- Also fix improper logic for visiting extend bound wildcard and super bound wildcard. 

*** TODO ***
Add below test case when integrate the Z3 encoding for dataflow system. (Below test case would cause unsatisfiable solution in dataflow system with Z3 solver).
 
Test code:
```java
class C
{
    void m(Class<?> co) throws Exception {
        Class<?> o = co.forName("Integer");
    }
}
```

### CODE REVIEW

File: src/checkers/inference/VariableAnnotator.java [hunk 1]
```diff
@@ -1299,17 +1306,17 @@ public Void visitWildcard(AnnotatedWildcardType wildcardType, Tree tree) {
         final WildcardTree wildcardTree = (WildcardTree) tree;
         final Tree.Kind wildcardKind = wildcardTree.getKind();
         if (wildcardKind == Tree.Kind.UNBOUNDED_WILDCARD) {
-            addPrimaryVariable(wildcardType, tree);
-
-            //TODO: Add missing extends bounds
+            ArtificialExtendsBoundTree artificialExtendsBoundTree = artificialTreeBuilder.createArtificialExtendsBoundTree(wildcardTree);
```
Comment (topnessman) [inline src/checkers/inference/VariableAnnotator.java]: move type cast into constructor

File: src/checkers/inference/VariableAnnotator.java [hunk 2]
```diff
@@ -161,6 +167,7 @@ public VariableAnnotator(final InferenceAnnotatedTypeFactory typeFactory,
         this.classDeclAnnos = new HashMap<>();
         this.realChecker = realChecker;
         this.constraintManager = constraintManager;
+        this.artificialTreeBuilder = new ArtificialTreeBuilder((JavacProcessingEnvironment) inferenceTypeFactory.getProcessingEnv());
```
Comment (topnessman) [inline src/checkers/inference/VariableAnnotator.java]: move type cast into constructor of atb

File: src/checkers/inference/model/tree/ArtificialTreeBuilder.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+public class ArtificialTreeBuilder {
+
+    private final JavacProcessingEnvironment processingEnvironment;
+
+    public ArtificialTreeBuilder(JavacProcessingEnvironment processingEnvironment) {
+        this.processingEnvironment = processingEnvironment;
+    }
+
+    public ArtificialExtendsBoundTree createArtificialExtendsBoundTree(WildcardTree wildcardTree) {
+        Name objectName = Names.instance(processingEnvironment.getContext()).java_lang_Object;
```
Comment (topnessman) [inline src/checkers/inference/model/tree/ArtificialTreeBuilder.java]: new Name("java.lang.Object");

File: src/checkers/inference/model/tree/ArtificialExtendsBoundTree.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import com.sun.source.tree.WildcardTree;
+import com.sun.tools.javac.tree.JCTree.JCIdent;
+import com.sun.tools.javac.util.Name;
+
+public class ArtificialExtendsBoundTree extends JCIdent {
+
+    private final WildcardTree boundedWildcard;
+
+    public ArtificialExtendsBoundTree(Name name, WildcardTree wildcardTree) {
+        super(name, null);
```
Comment (topnessman) [inline src/checkers/inference/model/tree/ArtificialExtendsBoundTree.java]: Get rid of name, hard code name to be "x.x.ArtificialObject"

File: src/checkers/inference/model/tree/ArtificialTreeBuilder.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
+package checkers.inference.model.tree;
+
+import com.sun.source.tree.WildcardTree;
+import com.sun.tools.javac.processing.JavacProcessingEnvironment;
+import com.sun.tools.javac.util.Name;
+import com.sun.tools.javac.util.Names;
+
+public class ArtificialTreeBuilder {
```
Comment (topnessman) [inline src/checkers/inference/model/tree/ArtificialTreeBuilder.java]: Remove builder

File: src/checkers/inference/model/tree/ArtificialExtendsBoundTree.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        return (obj instanceof ArtificialExtendsBoundTree) &&
+                (this.boundedWildcard.equals(((ArtificialExtendsBoundTree) obj).getBoundedWildcard()));
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (topnessman) [inline src/checkers/inference/model/tree/ArtificialExtendsBoundTree.java]: Use hashcode utils , and pass wildcard tree

File: src/checkers/inference/VariableAnnotator.java [hunk 7]
```diff
@@ -1299,17 +1299,17 @@ public Void visitWildcard(AnnotatedWildcardType wildcardType, Tree tree) {
         final WildcardTree wildcardTree = (WildcardTree) tree;
         final Tree.Kind wildcardKind = wildcardTree.getKind();
         if (wildcardKind == Tree.Kind.UNBOUNDED_WILDCARD) {
-            addPrimaryVariable(wildcardType, tree);
-
-            //TODO: Add missing extends bounds
+            ArtificialExtendsBoundTree artificialExtendsBoundTree = new ArtificialExtendsBoundTree(wildcardTree);
+            addPrimaryVariable(wildcardType.getSuperBound(), tree);
+            addPrimaryVariable(wildcardType.getExtendsBound(), artificialExtendsBoundTree);
```
Comment (CharlesZ-Chen) [inline src/checkers/inference/VariableAnnotator.java]: Change order.

File: src/checkers/inference/model/tree/ArtificialExtendsBoundTree.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
+package checkers.inference.model.tree;
+
+import org.checkerframework.dataflow.util.HashCodeUtils;
+
+import com.sun.source.tree.WildcardTree;
+import com.sun.tools.javac.tree.JCTree.JCIdent;
+
+public class ArtificialExtendsBoundTree extends JCIdent {
```
Comment (CharlesZ-Chen) [inline src/checkers/inference/model/tree/ArtificialExtendsBoundTree.java]: Add javadoc.

File: src/checkers/inference/model/tree/ArtificialExtendsBoundTree.java (hunk 6 above)
Comment (CharlesZ-Chen) [reply src/checkers/inference/model/tree/ArtificialExtendsBoundTree.java]: Done.

**Pull request merged**